cli/src/commands/bookmark/list.rs

```diff
@@ -277,3 +307,492 @@ struct RefListItem {
+    // information.
+    fn prepare_data_sort_and_snapshot(sort_keys: &[SortKey]) -> String {
+        let mut new_commit_id = commit_id_generator();
+        let mut new_timestamp = commit_ts_generator();
+        let names = ["bob", "alice", "eve", "bob"];
+        let emails = ["bob@g.com", "alice@g.com", "eve@g.com", "bob@g.com"];
+        let bookmark_names = ["feature", "bug-fix", "chore", "bug-fix"];
+        let remote_names = [None, Some("upstream"), None, Some("origin")];
+        let mut bookmark_items: Vec<RefListItem> = Vec::new();
+        let mut commits: HashMap<CommitId, Arc<backend::Commit>> = HashMap::new();
+        for (((&name, &email), bookmark_name), remote_name) in names
+            .iter()
+            .zip(emails.iter())
+            .zip(bookmark_names.iter())
+            .zip(remote_names.iter())
+        {
+            let commit_id = new_commit_id();
+            let mut b_name = "foo";
+            let mut author = make_default_signature();
+            let mut committer = make_default_signature();
+
+            if sort_keys.contains(&SortKey::Name) || sort_keys.contains(&SortKey::NameDesc) {
+                b_name = bookmark_name;
+            }
+            if sort_keys.contains(&SortKey::AuthorName)
+                || sort_keys.contains(&SortKey::AuthorNameDesc)
+            {
+                author.name = String::from(name);
+            }
+            if sort_keys.contains(&SortKey::AuthorEmail)
+                || sort_keys.contains(&SortKey::AuthorEmailDesc)
+            {
+                author.email = String::from(email);
+            }
+            if sort_keys.contains(&SortKey::AuthorDate)
+                || sort_keys.contains(&SortKey::AuthorDateDesc)
+            {
+                author.timestamp = new_timestamp();
+            }
+            if sort_keys.contains(&SortKey::CommitterName)
+                || sort_keys.contains(&SortKey::CommitterNameDesc)
+            {
+                committer.name = String::from(name);
+            }
+            if sort_keys.contains(&SortKey::CommitterEmail)
+                || sort_keys.contains(&SortKey::CommitterEmailDesc)
+            {
+                committer.email = String::from(email);
+            }
+            if sort_keys.contains(&SortKey::CommitterDate)
+                || sort_keys.contains(&SortKey::CommitterDateDesc)
+            {
+                committer.timestamp = new_timestamp();
+            }
+
+            if let Some(remote_name) = remote_name {
+                let local_target = RefTarget::normal(commit_id.clone());
+                let remote_ref = RemoteRef {
+                    target: local_target.clone(),
+                    state: RemoteRefState::New,
+                };
+                bookmark_items.push(RefListItem {
+                    primary: CommitRef::remote(b_name, *remote_name, remote_ref, &local_target),
+                    tracked: vec![],
+                });
```

Comment:
nit: should be `CommitRef::remote_only()` (it wouldn't matter in tests, though.)